File: prisma/inspect_prisma_duplicates.py
import os
import re
from collections import defaultdict

os.chdir(r"c:\Users\simoo\OneDrive\Bureau\SIMOCASINO\prisma")
models = {}
current = None
re_model = re.compile(r'^model\s+(\w+)\s*\{')

for fname in os.listdir('.'):
    if not fname.endswith('.prisma'):
        continue
    with open(fname, 'r', encoding='utf-8') as f:
        for line in f:
            line = line.rstrip('\n')
            m = re_model.match(line)
            if m:
                current = m.group(1)
                models.setdefault(current, [])
            elif current and line.strip() and not line.strip().startswith('//') and not line.strip().startswith('@@'):
                fm = re.match(r'^[A-Za-z_][A-Za-z0-9_]*\s+', line.strip())
                if fm:
                    models[current].append(fm.group(0).strip())
            if line.strip() == '}':
                current = None

print('DUPLICATE FIELDS WITHIN MODELS')
for model, fields in models.items():
    seen = set()
    dup = []
    for f in fields:
        if f in seen and f not in dup:
            dup.append(f)
        else:
            seen.add(f)
    if dup:
        print(model, dup)

field_use = defaultdict(set)
for model, fields in models.items():
    for f in fields:
        field_use[f].add(model)

print('\nFIELDS APPEARING IN MANY MODELS (>=4)')
for f, ms in sorted(field_use.items(), key=lambda x: (-len(x[1]), x[0])):
    if len(ms) >= 4:
        print(f, len(ms), sorted(ms))

File: prisma/inspect_prisma_duplicates.js
const fs = require('fs');
const path = require('path');
const re = /^model\s+(\w+)\s*\{/;
const models = {};
let current = null;
for (const fname of fs.readdirSync('.')) {
  if (!fname.endsWith('.prisma')) continue;
  const text = fs.readFileSync(fname, 'utf8');
  for (const line of text.split(/\r?\n/)) {
    const m = line.match(re);
    if (m) {
      current = m[1];
      models[current] = models[current] || [];
    } else if (current && line.trim() && !line.trim().startsWith('//') && !line.trim().startsWith('@@')) {
      const fm = line.trim().match(/^[A-Za-z_][A-Za-z0-9_]*\s+/);
      if (fm) models[current].push(fm[0].trim());
    }
    if (line.trim() === '}') current = null;
  }
}
console.log('DUPLICATE FIELDS WITHIN MODELS');
for (const [model, fields] of Object.entries(models)) {
  const seen = new Set();
  const dup = [];
  for (const f of fields) {
    if (seen.has(f) && !dup.includes(f)) dup.push(f);
    else seen.add(f);
  }
  if (dup.length) console.log(model, dup);
}
const fieldUse = {};
for (const [model, fields] of Object.entries(models)) {
  for (const f of fields) {
    if (!fieldUse[f]) fieldUse[f] = new Set();
    fieldUse[f].add(model);
  }
}
console.log('\nFIELDS APPEARING IN MANY MODELS (>=4)');
Object.entries(fieldUse)
  .sort((a, b) => b[1].size - a[1].size || a[0].localeCompare(b[0]))
  .forEach(([f, ms]) => {
    if (ms.size >= 4) console.log(f, ms.size, [...ms].sort());
  });
